plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs

```diff
@@ -316,40 +316,22 @@ flattenValue v = goOuter [] (Map.toList $ getValue v)
 isZero :: Value -> Bool
 isZero (Value xs) = Map.all (Map.all (\i -> 0 == i)) xs
 
-{-# INLINABLE checkPred #-}
-checkPred :: (These Integer Integer -> Bool) -> Value -> Value -> Bool
-checkPred f l r =
-    let
-      inner :: Map.Map TokenName (These Integer Integer) -> Bool
-      inner = Map.all f
-    in
-      Map.all inner (unionVal l r)
-
-{-# INLINABLE checkBinRel #-}
--- | Check whether a binary relation holds for value pairs of two 'Value' maps,
---   supplying 0 where a key is only present in one of them.
-checkBinRel :: (Integer -> Integer -> Bool) -> Value -> Value -> Bool
-checkBinRel f l r =
-    let
-        unThese k' = case k' of
-            This a    -> f a 0
-            That b    -> f 0 b
-            These a b -> f a b
-    in checkPred unThese l r
-
 {-# INLINABLE geq #-}
 -- | Check whether one 'Value' is greater than or equal to another. See 'Value' for an explanation
 -- of how operations on 'Value's work.
 geq :: Value -> Value -> Bool
--- If both are zero then checkBinRel will be vacuously true, but this is fine.
-geq = checkBinRel (>=)
+geq l (Value r) =
+  all
```

Comment:
Maybe define `valueEntries :: Value -> [(CurrencyId, TokenId, Integer)]` function? Useful here and probably elsewhere.

Reply:
It already exists: `flattenValue`, but flattening both values before checking `geq` is more expensive than this.

Reply:
Aren't you pretty much exactly flattening the second value? I guess you don't construct an intermediary list this way. Perhaps instead we want `allValueEntries :: Value -> (CurrencyId, TokenId, Integer) -> Bool`. Or I guess the direct way is fine too.

Reply:
At least worth a comment explaining why you didn't use the existing function so I don't just refactor it back again later ;)

Reply:
> Aren't you pretty much exactly flattening the second value?

There's no need to flatten the _entire_ second value. We can short circuit.

Reply:
I guess this is the limitation with strict languages - it's much harder to fuse transformations without materializing intermediate results.

Reply:
> There's no need to flatten the entire second value. We can short circuit.

Good point. Comment!